alg/viewshed.cpp

```diff
@@ -426,499 +480,389 @@ bool Viewshed::run(GDALRasterBandH hBand, GDALProgressFunc pfnProgress,
+    // Make the current line the last line.
+    vLastLineVal = std::move(vThisLineVal);
 
-            setVisibility(iPixel, dfZ, padfFirstLineVal, vResult);
-        }
-        else
-        {
-            for (; iPixel < nXSize; iPixel++)
-            {
-                pabyResult[iPixel] = oOpts.outOfRangeVal;
-                if (oOpts.outputMode != OutputMode::Normal)
-                    dfHeightResult[iPixel] = oOpts.outOfRangeVal;
-            }
-        }
-    }
-    /* write result line */
+    // Create the output writer.
+    if (!writeLine(nY, vResult))
+        return false;
 
-    void *data;
-    GDALDataType dataType;
-    if (oOpts.outputMode == OutputMode::Normal)
+    if (!lineProgress())
+        return false;
+    return true;
+}
+
+/// Process a line above or below the observer.
+///
+/// @param nX  X location of the observer
+/// @param nY  Y location of the observer
+/// @param nLine  Line number being processed.
+/// @param vLastLineVal  Vector in which to store the read line. Becomes the last line
+///    in further processing.
+/// @return True on success, false otherwise.
+bool Viewshed::processLine(int nX, int nY, int nLine,
+                           std::vector<double> &vLastLineVal)
+{
+    int nYOffset = nLine - nY;
+    std::vector<double> vResult(oOutExtent.xSize());
+    std::vector<double> vThisLineVal(oOutExtent.xSize());
+
+    if (!readLine(nLine, vThisLineVal.data()))
+        return false;
+
+    // In DEM mode the base is the input DEM value.
+    // In ground mode the base is zero.
+    if (oOpts.outputMode == OutputMode::DEM)
+        vResult = vThisLineVal;
+
+    // Adjust height of the read line.
+    const auto [iLeft, iRight] =
+        adjustHeight(nYOffset, nX, vThisLineVal.data() + nX);
+
+    // Handle the initial position on the line.
+    if (iLeft < iRight)
     {
-        data = static_cast<void *>(pabyResult);
-        dataType = GDT_Byte;
+        double dfZ = CalcHeightLine(nYOffset, vLastLineVal[nX]);
+        setOutput(vResult[nX], vThisLineVal[nX], dfZ);
     }
     else
-    {
-        data = static_cast<void *>(dfHeightResult);
-        dataType = GDT_Float64;
-    }
-    if (GDALRasterIO(hTargetBand, GF_Write, 0, nY - nYStart, nXSize, 1, data,
-                     nXSize, 1, dataType, 0, 0))
-    {
-        CPLError(CE_Failure, CPLE_AppDefined,
-                 "RasterIO error when writing target raster at position "
-                 "(%d,%d), size (%d,%d)",
-                 0, nY - nYStart, nXSize, 1);
-        return false;
-    }
+        vResult[nX] = oOpts.outOfRangeVal;
 
-    /* scan upwards */
-    std::copy(vFirstLineVal.begin(), vFirstLineVal.end(), vLastLineVal.begin());
-    for (int iLine = nY - 1; iLine >= nYStart; iLine--)
-    {
-        if (GDALRasterIO(hBand, GF_Read, nXStart, iLine, nXSize, 1,
-                         padfThisLineVal, nXSize, 1, GDT_Float64, 0, 0))
+    // process left half then right half of line
```

Comment:
I would be worried that spawning 2 threads, presumably very short lived unless you have gigantic rasters (I'm not even sure that raster width == INT_MAX would be big enough), repeatedly for each line might be more costly than the computation benefit they bring.

Reply:
If they're small rasters, why would anyone much care?